driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionCorpusTest.java

```diff
@@ -0,0 +1,292 @@
+                } else if (kms.equals("local")) {
+                    opts.keyId(new BsonBinary(BsonBinarySubType.UUID_STANDARD, localKeyId));
+                } else {
+                    throw new UnsupportedOperationException("Unsupported provider: " + kms);
+                }
+            } else if (identifier.equals("altname")) {
+                opts.keyAltName(kms);
+            } else {
+                throw new UnsupportedOperationException("Unsupported identifier: " + identifier);
+            }
+
+            try {
+                BsonValue encryptedValue = clientEncryption.encrypt(value, opts);
+                fieldDocument.put("value", encryptedValue);
+                corpusCopied.append(field, fieldDocument);
+            } catch (Exception e) {
+                assertFalse(allowed);
+                corpusCopied.append(field, fieldDocument);
+            }
+        }
+
+        // Step 6: insert corpusCopied
+        MongoCollection<BsonDocument> encryptedCollection = autoEncryptingClient.getDatabase("db")
+                .getCollection("coll", BsonDocument.class);
+        encryptedCollection.insertOne(corpusCopied);
+
+        // Step 7: check the auto decrypted document
+        BsonDocument corpusDecrypted = encryptedCollection.find(new BsonDocument()).first();
+        assertEquals(corpus, corpusDecrypted);
+
+        // Step 8: check the document with an unencrypted client
+        MongoCollection coll = client.getDatabase("db").getCollection("coll", BsonDocument.class);
+        BsonDocument corpusEncryptedActual = (BsonDocument) coll.find(new BsonDocument()).first();
+        BsonDocument corpusEncryptedExpected = bsonDocumentFromPath("corpus-encrypted.json");
+
+        for (String field : corpusEncryptedActual.keySet()) {
+            if (field.equals("_id") || field.equals("altname_aws") || field.equals("altname_local")) {
+                continue;
+            }
+
+            String kms = corpusEncryptedActual.getDocument(field).getString("kms").getValue();
+            String type = corpusEncryptedActual.getDocument(field).getString("type").getValue();
+            String algo = corpusEncryptedActual.getDocument(field).getString("algo").getValue();
+            String method = corpusEncryptedActual.getDocument(field).getString("method").getValue();
+            // TODO: unused
+            String identifier = corpusEncryptedActual.getDocument(field).getString("identifier").getValue();
+            boolean allowed = corpusEncryptedActual.getDocument(field).getBoolean("allowed").getValue();
+            BsonValue value = corpusEncryptedActual.getDocument(field).get("value");
+
+            // All deterministic fields are an exact match.
+            if (algo.equals("det")) {
+                deterministicCheck(kms, type, corpusEncryptedExpected);
+            }
+
+            if (algo.equals("rand") && !method.equals("prohibited")) {
```

Comment:
Ah, this is my mistake. After the values were turned into documents, the `method` was changed to be only `explicit` and `auto`. Instead of `!method.equals("prohibited")` this should be `allowed`.

I now think `randomCheck` is wrong, as this should have been failing on disallowed fields. See my comment below.

Reply:
Done